--- Makefile-ref	2011-02-28 23:04:15.000000000 +0300
+++ Makefile	2018-02-21 16:07:35.985012239 +0300
@@ -2,6 +2,9 @@
 #          Compression build options          #
 ###############################################
 #
+SRCBASE ?= $(shell pwd)/..
+CC = gcc
+export CC
 #
 ############# Building gzip support ###########
 #
@@ -26,7 +29,9 @@
 # To build using XZ Utils liblzma - install the library and uncomment
 # the XZ_SUPPORT line below.
 #
-#XZ_SUPPORT = 1
+XZ_SUPPORT = 1
+#XZUTILSDIR = $(SRCBASE)/xz
+XZUTILSDIR = $(SRCBASE)/xz-5.2.4
 
 
 ############ Building LZO support ##############
@@ -67,7 +72,7 @@
 # in Mksquashfs.  Obviously the compression algorithm must have been
 # selected to be built
 #
-COMP_DEFAULT = gzip
+COMP_DEFAULT = xz
 
 ###############################################
 #  Extended attribute (XATTRs) build options  #
@@ -93,8 +98,8 @@
 #        End of BUILD options section         #
 ###############################################
 
-INCLUDEDIR = -I.
-INSTALL_DIR = /usr/local/bin
+INCLUDEDIR = -I. -I/usr
+INSTALL_DIR = $(SRCBASE)/bin
 
 MKSQUASHFS_OBJS = mksquashfs.o read_fs.o sort.o swap.o pseudo.o compressor.o
 
@@ -135,9 +140,14 @@
 
 ifeq ($(XZ_SUPPORT),1)
 CFLAGS += -DXZ_SUPPORT
+ifdef XZUTILSDIR
+INCLUDEDIR += -I$(XZUTILSDIR)/src/liblzma/api
+XZ_LIBDIR = -L$(XZUTILSDIR)/src/liblzma/.libs
+LIBS += $(XZUTILSDIR)/src/liblzma/.libs/liblzma.a
+endif
 MKSQUASHFS_OBJS += xz_wrapper.o
 UNSQUASHFS_OBJS += xz_wrapper.o
-LIBS += -llzma
+#LIBS += $(XZ_LIBDIR) -llzma
 COMPRESSORS += xz
 endif
 
@@ -196,9 +206,9 @@
 endif
 
 .PHONY: all
-all: mksquashfs unsquashfs
+all: xzutils mksquashfs unsquashfs
 
-mksquashfs: $(MKSQUASHFS_OBJS)
+mksquashfs: xzutils $(MKSQUASHFS_OBJS)
 	$(CC) $(LDFLAGS) $(EXTRA_LDFLAGS) $(MKSQUASHFS_OBJS) $(LIBS) -o $@
 
 mksquashfs.o: mksquashfs.c squashfs_fs.h mksquashfs.h sort.h squashfs_swap.h \
@@ -254,5 +264,24 @@
 .PHONY: install
 install: mksquashfs unsquashfs
 	mkdir -p $(INSTALL_DIR)
-	cp mksquashfs $(INSTALL_DIR)
-	cp unsquashfs $(INSTALL_DIR)
+	cp mksquashfs $(INSTALL_DIR)/mksquashfs_xz-4.0
+	cp unsquashfs $(INSTALL_DIR)/unsquashfs_xz-4.0
+
+.PHONY: xzutils xzutils-configure
+#xzutils: $(XZUTILSDIR)/configure $(XZUTILSDIR)/Makefile
+xzutils: xzutils-autogen xzutils-Makefile
+	$(MAKE) -C $(XZUTILSDIR)
+
+#$(XZUTILSDIR)/Makefile:
+xzutils-Makefile:
+	$(MAKE) xzutils-configure
+
+#$(XZUTILSDIR)/configure:
+xzutils-autogen:
+	( cd $(XZUTILSDIR) ; ./autogen.sh )
+
+xzutils-configure:
+	( cd $(XZUTILSDIR) ; \
+		./configure \
+		--prefix=/usr/local \
+	)
