diff -uNrwB xl2tpd-w/CHANGES xl2tpd/CHANGES
--- xl2tpd-w/CHANGES	2018-04-20 20:34:13.057022021 +0600
+++ xl2tpd/CHANGES	2018-04-20 20:34:16.568278392 +0600
@@ -1,3 +1,10 @@
+v1.3.11 (March 7, 2018)
+* Build packages for Xenial by default (Simon Deziel)
+* Bump d/compat to 9 (Simon Deziel)
+* Drop d/repack.sh script and refresh d/watch (Simon Deziel)
+* Refresh d/control by partly sync'ing from Debian (Simon Deziel)
+* Use HTTPS URL in d/copyright (Simon Deziel)
+
 v1.3.10.1 (November 16, 2017)
 * Have max retries as a configuration [Samir Hussain]
 * Add more into to "select timeout" debug message [Samir Hussain]
diff -uNrwB xl2tpd-w/debian/changelog xl2tpd/debian/changelog
--- xl2tpd-w/debian/changelog	2018-04-20 20:34:13.069022897 +0600
+++ xl2tpd/debian/changelog	2018-04-20 20:34:16.569278465 +0600
@@ -1,4 +1,4 @@
-xl2tpd (XL2TPDBASEVERSION-0xelerance1) trusty; urgency=low
+xl2tpd (XL2TPDBASEVERSION-0ubuntu0.16.04xelerance1) xenial; urgency=low
 
   * Update to XL2TPDBASEVERSION
 
diff -uNrwB xl2tpd-w/debian/compat xl2tpd/debian/compat
--- xl2tpd-w/debian/compat	2018-04-20 20:34:13.070022970 +0600
+++ xl2tpd/debian/compat	2018-04-20 20:34:16.569278465 +0600
@@ -1 +1 @@
-7
+9
diff -uNrwB xl2tpd-w/debian/control xl2tpd/debian/control
--- xl2tpd-w/debian/control	2018-04-20 20:34:13.070022970 +0600
+++ xl2tpd/debian/control	2018-04-20 20:34:16.569278465 +0600
@@ -5,13 +5,13 @@
 Homepage: https://www.xelerance.com/software/xl2tpd/
 Vcs-Browser: https://github.com/xelerance/xl2tpd
 Vcs-Git: git://github.com/xelerance/xl2tpd.git
-Build-Depends: debhelper (>= 7.0.8), libpcap0.8-dev
-Standards-Version: 3.9.5
+Build-Depends: debhelper (>= 9.20120115), libpcap0.8-dev
+Standards-Version: 3.9.8
 
 Package: xl2tpd
 Architecture: any
 Provides: l2tpd
-Depends: ${shlibs:Depends}, ${misc:Depends}, ppp
+Depends: ${shlibs:Depends}, ${misc:Depends}, ppp, lsb-base (>= 3.0-6)
 Description: layer 2 tunneling protocol implementation
  xl2tpd is an open source implementation of the L2TP tunneling
  protocol (RFC2661).  xl2tpd is forked from l2tpd and is maintained by
diff -uNrwB xl2tpd-w/debian/copyright xl2tpd/debian/copyright
--- xl2tpd-w/debian/copyright	2018-04-20 20:34:13.071023043 +0600
+++ xl2tpd/debian/copyright	2018-04-20 20:34:16.569278465 +0600
@@ -1,4 +1,4 @@
-Format: http://www.debian.org/doc/packaging-manuals/copyright-format/1.0/
+Format: https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/
 Source: https://github.com/xelerance/xl2tpd
 
 Files: *
diff -uNrwB xl2tpd-w/debian/repack.sh xl2tpd/debian/repack.sh
--- xl2tpd-w/debian/repack.sh	2018-04-20 20:34:13.071023043 +0600
+++ xl2tpd/debian/repack.sh	1970-01-01 06:00:00.000000000 +0600
@@ -1,41 +0,0 @@
-#!/bin/sh
-# Repackage upstream source to exclude non-distributable files
-# should be called as "repack sh --upstream-source <ver> <downloaded file>
-# (for example, via uscan)
-
-set -e
-set -u
-
-FILE=$3
-PKG=`dpkg-parsechangelog|grep ^Source:|sed 's/^Source: //'`
-VER="$2+dfsg"
-
-printf "\nRepackaging $FILE\n"
-
-DIR=`mktemp -d ./tmpRepackXXXXXX`
-trap "rm -rf $DIR" QUIT INT EXIT
-
-tar xzf $FILE -C $DIR
-
-TARGET=`echo $FILE |sed 's/_\(.*\)\.orig/_\1+dfsg.orig/'`
-REPACK=`basename $TARGET`
-
-UP_DIR=`ls -1 $DIR`
-
-(
-    set -e
-    set -u
-
-    cd $DIR
-
-    rm -rv $UP_DIR/debian/
-
-    REPACK_DIR="$PKG-$VER.orig"
-    mv $UP_DIR $REPACK_DIR
-    tar -c $REPACK_DIR | gzip -9 > $REPACK
-)
-
-rm -v $FILE
-mv $DIR/$REPACK $TARGET
-
-echo "*** $FILE repackaged as $TARGET"
diff -uNrwB xl2tpd-w/debian/watch xl2tpd/debian/watch
--- xl2tpd-w/debian/watch	2018-04-20 20:34:13.073023189 +0600
+++ xl2tpd/debian/watch	2018-04-20 20:34:16.570278538 +0600
@@ -1,9 +1,3 @@
-# The uversionmangle option separates consecutive digits in the version string
-# with a period.  The reason for this is that in the past, upstream used
-# version numbers like 1.00, 1.00, and 1.1.08.  For some time now they have
-# switched to a more "standard" 1.2.8, 1.3.0, 1.3.1, and so forth, versioning
-# scheme.
-
-version=3
-opts="dversionmangle=s/\+dfsg//,uversionmangle=s/(\d)(\d)/\1\.\2/g" https://github.com/xelerance/xl2tpd/tags .*/v?(\d[\d\.]+)\.(?:tgz|tbz2|txz|tar\.(?:gz|bz2|xz)) debian debian/repack.sh
-
+version=4
+opts=filenamemangle=s/.+\/v?(\d\S+)\.tar\.gz/xl2tpd-$1\.tar\.gz/ \
+  https://github.com/xelerance/xl2tpd/tags .*/v?(\d\S+)\.tar\.gz
diff -uNrwB xl2tpd-w/l2tp.h xl2tpd/l2tp.h
--- xl2tpd-w/l2tp.h	2018-04-20 20:34:13.093024649 +0600
+++ xl2tpd/l2tp.h	2018-04-20 20:34:16.570278538 +0600
@@ -58,7 +58,7 @@
 #define CONTROL_PIPE_REQ_LNS_REMOVE 'w'     /* Get status of LNS */
 
 #define BINARY "xl2tpd"
-#define SERVER_VERSION "xl2tpd-1.3.10.1"
+#define SERVER_VERSION "xl2tpd-1.3.11"
 #define VENDOR_NAME "xelerance.com"
 #ifndef PPPD
 #define PPPD		"/usr/sbin/pppd"
diff -uNrwB xl2tpd-w/packaging/fedora/xl2tpd.spec xl2tpd/packaging/fedora/xl2tpd.spec
--- xl2tpd-w/packaging/fedora/xl2tpd.spec	2018-04-20 20:34:13.098025014 +0600
+++ xl2tpd/packaging/fedora/xl2tpd.spec	2018-04-20 20:34:16.571278611 +0600
@@ -1,6 +1,6 @@
 Summary: Layer 2 Tunnelling Protocol Daemon (RFC 2661)
 Name: xl2tpd
-Version: 1.3.10.1
+Version: 1.3.11
 Release: 1%{?dist}
 License: GPLv2
 Url: http://www.xelerance.com/software/xl2tpd/
diff -uNrwB xl2tpd-w/packaging/openwrt/Makefile xl2tpd/packaging/openwrt/Makefile
--- xl2tpd-w/packaging/openwrt/Makefile	2018-04-20 20:34:13.099025087 +0600
+++ xl2tpd/packaging/openwrt/Makefile	2018-04-20 20:34:16.571278611 +0600
@@ -2,7 +2,7 @@
 include $(TOPDIR)/rules.mk
 
 PKG_NAME:=xl2tpd
-PKG_VERSION:=1.3.10.1
+PKG_VERSION:=1.3.11
 PKG_RELEASE:=1
 PKG_MD5SUM:=ab5656eb5a3d1973f7f69b039675332e-NEEDSUPDATING
 
diff -uNrwB xl2tpd-w/packaging/suse/sles10.spec xl2tpd/packaging/suse/sles10.spec
--- xl2tpd-w/packaging/suse/sles10.spec	2018-04-20 20:34:13.102025306 +0600
+++ xl2tpd/packaging/suse/sles10.spec	2018-04-20 20:34:16.572278684 +0600
@@ -1,6 +1,6 @@
 Summary: Layer 2 Tunnelling Protocol Daemon (RFC 2661)
 Name: xl2tpd
-Version: 1.3.10.1
+Version: 1.3.11
 Release: 1%{?dist}
 License: GPLv2
 Url: http://www.xelerance.com/software/xl2tpd/
diff -uNrwB xl2tpd-w/packaging/suse/xl2tpd.spec xl2tpd/packaging/suse/xl2tpd.spec
--- xl2tpd-w/packaging/suse/xl2tpd.spec	2018-04-20 20:34:13.103025379 +0600
+++ xl2tpd/packaging/suse/xl2tpd.spec	2018-04-20 20:34:16.573278757 +0600
@@ -1,6 +1,6 @@
 Summary: Layer 2 Tunnelling Protocol Daemon (RFC 2661)
 Name: xl2tpd
-Version: 1.3.10.1
+Version: 1.3.11
 Release: 1%{?dist}
 License: GPLv2
 Url: http://www.xelerance.com/software/xl2tpd/
diff -uNrwB xl2tpd-w/README.xl2tpd xl2tpd/README.xl2tpd
--- xl2tpd-w/README.xl2tpd	2018-04-20 20:34:13.059022167 +0600
+++ xl2tpd/README.xl2tpd	2018-04-20 20:34:16.568278392 +0600
@@ -26,6 +27,11 @@
 Xl2tpd is based on the 0.69 L2TP by Jeff McAdams <jeffm@iglou.com>
 It was de-facto maintained by Jacco de Leeuw <jacco2@dds.nl> in 2002 and 2003.
 
+NOTE: In Linux kernel 4.15+ there is a kernel bug with ancillary IP_PKTINFO.
+      As such, for Linux kernel 4.15+ we recommend the community use the
+      1.3.12dev branch as a temporary measure until the branch is merged into
+      the master branch
+
 Mailing Lists :
     https://lists.openswan.org/cgi-bin/mailman/listinfo/xl2tpd is home of the
     mailing list. Note: This is a closed list - you *must* be subscribed to post.
