--- coova-chilli-1.3.0/Makefile.am	2015-08-10 14:08:38.978483397 +0600
+++ coova-chilli-master/Makefile.am	2015-08-19 22:14:03.000000000 +0600
@@ -5,8 +5,7 @@
 if WITH_JSONLIB
 SUBDIRS += json
 endif
-SUBDIRS += src www conf
-# doc debian distro
+SUBDIRS += src doc www conf debian distro
 if WITH_MINIPORTAL
 SUBDIRS += miniportal
 endif
--- coova-chilli-1.3.0/src/Makefile.am	2015-08-13 04:42:33.762838360 +0600
+++ coova-chilli-master/src/Makefile.am	2015-08-19 22:14:03.000000000 +0600
@@ -25,12 +25,12 @@
 iphash.c lookup.c system.h util.c options.c statusfile.c conn.c sig.c \
 garden.c dns.c session.c pkt.c chksum.c net.c safe.c
 
-AM_CFLAGS = -D_GNU_SOURCE -Wall -fno-builtin -fno-strict-aliasing \
+AM_CFLAGS = -D_GNU_SOURCE -Wall -Werror -fno-builtin -fno-strict-aliasing \
   -fomit-frame-pointer -funroll-loops -pipe -I$(top_builddir)/bstring \
- -DDEFCHILLICONF='"/etc/chilli.conf"'\
- -DDEFPIDFILE='"/var/run/chilli.pid"'\
- -DDEFSTATEDIR='"/var/run"'\
- -DSBINDIR='"/sbin"' 
+ -DDEFCHILLICONF='"$(sysconfdir)/chilli.conf"'\
+ -DDEFPIDFILE='"$(localstatedir)/run/chilli.pid"'\
+ -DDEFSTATEDIR='"$(localstatedir)/run"'\
+ -DSBINDIR='"$(sbindir)"' 
 
 chilli_SOURCES= main.c
 chilli_radconfig_SOURCES = main-radconfig.c 
@@ -194,9 +194,9 @@
 main-opt.c: cmdline.c cmdline.h
 
 cmdline.c cmdline.h: cmdline.ggo cmdline.patch
-#	cat $(CMDLINE) | gengetopt -C 
-#	cp cmdline.c cmdline.c.orig
-#	patch -p0 < cmdline.patch
+	cat $(CMDLINE) | gengetopt -C 
+	cp cmdline.c cmdline.c.orig
+	patch -p0 < cmdline.patch
 
 cmdline.mini.gen: cmdline.ggo 
 	sh cmdline.mini.sh > cmdline.mini.gen
